clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -110,7 +110,9 @@ public synchronized Cluster fetch() {
      * will be reset on the next update.
      */
     public synchronized void add(String topic) {
-        topics.put(topic, TOPIC_EXPIRY_NEEDS_UPDATE);
+        if (topics.put(topic, TOPIC_EXPIRY_NEEDS_UPDATE) == null) {
```

Comment:
Hmm, after I added 1a33308, I found that a test start failing.
```
org.apache.kafka.clients.consumer.KafkaConsumerTest > testChangingRegexSubscription FAILED
    java.lang.IllegalStateException: Next in line response did not match expected request
        at org.apache.kafka.clients.MockClient.send(MockClient.java:147)
        at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.trySend(ConsumerNetworkClient.java:392)
        at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:223)
        at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:180)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:308)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:277)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:259)
        at org.apache.kafka.clients.consumer.KafkaConsumer.pollOnce(KafkaConsumer.java:1012)
        at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:978)
        at org.apache.kafka.clients.consumer.KafkaConsumerTest.testChangingRegexSubscription(KafkaConsumerTest.java:623)
```
This is because the current regexp-based subscription relies on `Metadata.Listener` to update the list of topics that it subscribing on: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L176

So this line still isn't removable even though we apply 1a33308.
However it might still make sense to keep the change as it reduced few LOC in other situations and keep calling `requestUpdate()` explicitly for the cases like this.
I'm on the line to revert 1a33308 as I think it is making the responsibility of calling `requestUpdate()` among caller and Metadata but wanna hear your opinion @hachikuji .